### PR TITLE
feat(state-viewer): list actions invoked by each contract

### PR DESCRIPTION
**Pull request by jakmeier**
List account names with contracts deployed and additional 
information about the contracts.

See the state-viewer README for docs and examples.

This has been used to produce reports in #8427

### CODE REVIEW

**Comment by jakmeier** [comment]
@pugachAG maybe you can review this? It's an extention to the tool you alraedy reviewed.

**Review by pugachAG — APPROVED** [review]
Overall looks good to me!

**Review by akhi3030 — APPROVED** [review]
Approving to unblock merging.

File: tools/state-viewer/src/commands.rs [hunk 1]
```diff
@@ -852,11 +854,12 @@ pub(crate) fn contract_accounts(
     home_dir: &Path,
     store: Store,
     near_config: NearConfig,
+    filter: ContractAccountFilter,
 ) -> anyhow::Result<()> {
     let (_runtime, state_roots, _header) = load_trie(store.clone(), home_dir, &near_config);
 
-    for (shard_id, &state_root) in state_roots.iter().enumerate() {
-        eprintln!("Starting shard {shard_id}");
+    // shard_accounts.push(ContractAccount::in_trie(tries.last().unwrap())?);
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply tools/state-viewer/src/commands.rs]: removed, thanks

File: tools/state-viewer/src/commands.rs [hunk 2]
```diff
@@ -866,14 +869,31 @@ pub(crate) fn contract_accounts(
...
-
-        for contract in ContractAccount::in_trie(&trie)? {
-            match contract {
-                Ok(contract) => println!("{contract}"),
-                Err(err) => eprintln!("{err}"),
+        Trie::new(Box::new(storage), state_root, flat_state)
+    });
+
+    filter.write_header(&mut std::io::stdout().lock())?;
+    let streaming = false;
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply tools/state-viewer/src/commands.rs]: oh, that's missing an implementation :)
Added now. Can you please take another look? @pugachAG

File: tools/state-viewer/src/contract_accounts.rs [hunk 3]
```diff
@@ -1,65 +1,188 @@
...
-        let account_id = parse_account_id_from_contract_code_key(trie_key)
-            .map_err(|err| ContractAccountError::InvalidKey(err, trie_key.to_vec()))?;
-        let source_wasm = trie
-            .storage
-            .retrieve_raw_bytes(&value_hash)
-            .map_err(|err| ContractAccountError::NoCode(err, account_id.clone()))?;
-        Ok(Self { account_id, source_wasm })
+        filter: &ContractAccountFilter,
+    ) -> Result<Self> {
+        let get_code = || {
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply tools/state-viewer/src/contract_accounts.rs]: Right, inlining is better. 

(A previous iteration of the code also allowed to keep the entire code, which was a different field and so I was calling `get_code` it twice. Now that's no longer the case.)

File: tools/state-viewer/src/contract_accounts.rs [hunk 4]
```diff
@@ -78,36 +201,290 @@ impl<'a> ContractAccountIterator<'a> {
...
+    fn summary(self, store: &Store, filter: &ContractAccountFilter) -> ContractAccountSummary;
+}
+
+impl<T: Iterator<Item = Result<ContractAccount>>> Summary for T {
+    /// Eagerly evaluate everything selected by the filter, including fields
+    /// that are not available in the streaming iterator.
+    fn summary(self, store: &Store, filter: &ContractAccountFilter) -> ContractAccountSummary {
+        let mut errors = vec![];
+        let mut contracts: BTreeMap<_, _> = self
+            .flat_map(|result| match result {
```
[the remark this answers is not in the document]
Comment (jakmeier) [reply tools/state-viewer/src/contract_accounts.rs]: ah right, that's much nicer

**Review by pugachAG — APPROVED** [review]

**Pull request merged**